### PR TITLE
mimxrt10xx: add one-directional UART

### LINKED ISSUE #2484: mimxrt10xx: Allow one-way UART
Right now UART expects TX and RX, need to fix that.

### PR DESCRIPTION
**Pull request by hierophect**
This PR enables one-directional UART to Busio, allowing for the definition of a UART line that has only the TX or RX pin. 

Throws an error if you try to do this while also defining pins for flow control, since the two are basically incompatible as far as I'm aware. Also adds the reserved_uart table so the same peripheral instance can't be used twice, and fixes some bugs I found in the stm32 UART implementation while using it for reference (whoops). 

resolves #2484

### CODE REVIEW

**Comment by hierophect** [comment]
Going to follow this and #2898 up with a general Busio testing and cleanup PR that brings style in line across all three and enables proper pin resetting / never reset now that those functions are enabled.

File: locale/circuitpython.pot [hunk 1]
```diff
@@ -1518,6 +1522,10 @@ msgstr ""
 msgid "UART Re-init error"
 msgstr ""
 
+#: ports/mimxrt10xx/common-hal/busio/UART.c
+msgid "UART peripheral is already in use"
```
Comment (tannewt) [inline locale/circuitpython.pot:1526]: Please reuse error messages from other ports when it makes sense. That way there are less strings to translate. Here is what STM does: https://github.com/adafruit/circuitpython/blob/master/ports/stm/common-hal/busio/UART.c#L58

File: ports/mimxrt10xx/common-hal/busio/UART.c [hunk 2]
```diff
@@ -39,7 +39,9 @@
 
 #include "fsl_lpuart.h"
 
-// TODO
+//arrays use 0 based numbering: UART is stored at index 0
```
Comment (tannewt) [inline ports/mimxrt10xx/common-hal/busio/UART.c:42]: Do you mean:

```suggestion
//arrays use 0 based numbering: UART1 is stored at index 0
```

File: ports/mimxrt10xx/common-hal/busio/UART.c [hunk 3]
```diff
@@ -79,126 +81,180 @@ void common_hal_busio_uart_construct(busio_uart_obj_t *self,
...
-    }
-
     self->baudrate = baudrate;
     self->character_bits = bits;
     self->timeout_ms = timeout * 1000;
 
-    const uint32_t rx_count = sizeof(mcu_uart_rx_list) / sizeof(mcu_periph_obj_t);
-    const uint32_t tx_count = sizeof(mcu_uart_tx_list) / sizeof(mcu_periph_obj_t);
+    bool is_onedirection = false;
+    if (!rx != !tx) {
```
Comment (tannewt) [inline ports/mimxrt10xx/common-hal/busio/UART.c:89]: Please rewrite this check. It's not clear how it works.

File: ports/mimxrt10xx/common-hal/busio/UART.c [hunk 4]
```diff
@@ -229,16 +285,16 @@ void common_hal_busio_uart_deinit(busio_uart_obj_t *self) {
 
     gc_free(self->ringbuf);
 
-//    reset_pin_number(self->rx_pin);
-//    reset_pin_number(self->tx_pin);
+//    reset_pin_number(self->rx);
+//    reset_pin_number(self->tx);
```
Comment (tannewt) [inline ports/mimxrt10xx/common-hal/busio/UART.c:289]: Shouldn't these be uncommented?

File: ports/mimxrt10xx/common-hal/busio/UART.c (hunk 3 above)
Comment (hierophect) [reply ports/mimxrt10xx/common-hal/busio/UART.c:89]: It's an XOR, converts both to bool and compares to ensure they're different. Would a comment be OK or do you have another notation for xoring that you'd prefer?

File: ports/mimxrt10xx/common-hal/busio/UART.c (hunk 4 above)
Comment (hierophect) [reply ports/mimxrt10xx/common-hal/busio/UART.c:289]: I was going to do this in a follow up PR, since I wanted to keep this and the reset pins PR separate. I2C, SPI, DigitalIO and PWM all need them too.

**Comment by arturo182** [comment]
I have nothing more to add on top of Scott's comments :)

**Comment by hierophect** [comment]
@tannewt ready for another look.

File: ports/mimxrt10xx/common-hal/busio/UART.c (hunk 3 above)
Comment (tannewt) [reply ports/mimxrt10xx/common-hal/busio/UART.c]: I think the implicit bool conversion makes it more confusing and a comment would help. I'd suggest:

```c
// We are transmitting one direction if one pin is NULL and the other isn't.
bool is_onedirection = (rx != NULL) != (tx != NULL);
```

Comment (hierophect) [reply ports/mimxrt10xx/common-hal/busio/UART.c]: @tannewt yes that's much prettier, and good points about the implicit conversion. Nice solution!

**Comment by hierophect** [comment]
@tannewt all set, thanks for the suggestion

File: ports/mimxrt10xx/common-hal/busio/UART.c [hunk 5]
```diff
@@ -79,126 +81,178 @@ void common_hal_busio_uart_construct(busio_uart_obj_t *self,
...
+        mp_raise_ValueError(translate("Supply at least one UART pin"));
     }
 
-    if(self->rx_pin == NULL || self->tx_pin == NULL) {
+    if (uart_taken) {
+        mp_raise_RuntimeError(translate("Hardware in use, try alternative pins"));
+    }
+
+    if(self->rx == NULL && self->tx == NULL) {
         mp_raise_RuntimeError(translate("Invalid UART pin selection"));
```
Comment (arturo182) [inline ports/mimxrt10xx/common-hal/busio/UART.c:150]: This should probably be a separate PR, but in all the other ports this is a `mp_raise_ValueError(translate("Invalid pins"));`, it matters because it breaks, for example, this code: https://learn.adafruit.com/circuitpython-essentials/circuitpython-uart-serial#wheres-my-uart-2985036-14

**Comment by tannewt** [comment]
Oh, and @arturo182's suggesting of raising ValueError is good too. Please do that as well.

**Comment by hierophect** [comment]
@tannewt should be all set
